New GPIO input doc (#31599)
dotnet/docs: docs/iot/tutorials/gpio-input.md

General-purpose I/O (GPIO) pins can be configured to receive electrical signals as input. At its most basic level, this is useful for scenarios that detect the opening/closing of a circuit. Such circuits might include push buttons, toggle switches, reed switches, pressure switches, and other devices that represent binary (on/off) values by completing a circuit.

In this tutorial, you'll use .NET and your Raspberry Pi's GPIO pins to detect the opening and closing of a circuit.

## Prerequisites

- [!INCLUDE [prereq-rpi](../includes/prereq-rpi.md)]
- Jumper wires
- Breadboard (optional)
- Raspberry Pi GPIO breakout board (optional)
- [!INCLUDE [tutorial-prereq-dotnet](../includes/tutorial-prereq-dotnet.md)]

[!INCLUDE [ensure-ssh](../includes/ensure-ssh.md)]

## Prepare the hardware

Use the hardware components to build the circuit as depicted in the following diagram:

:::image type="content" source="https://via.placeholder.com/400x200.png" alt-text="A Fritzing diagram showing a circuit that connects a ground pin to pin 21." lightbox="https://via.placeholder.com/800x400.png":::

The image above depicts a direct connection between a ground pin and pin 21.

> [!TIP]

> > The diagram uses a breadboard and GPIO breakout for illustrative purposes, but feel free to just jumper a ground pin and pin 21 on the Raspberry Pi.

Reviewer: `"but feel free to just jumper a ground pin"`

I know what this is trying to communicate, but I'm not sure this phrasing is accurate. It's kind of hard to read, but I'm not sure how I'd rephrase it.

[!INCLUDE [tutorial-rpi-gpio](../includes/tutorial-rpi-gpio.md)]

## Create the app

Complete the following steps in your preferred development environment:

1. Create a new .NET Console App using either the [.NET CLI](../../core/tools/dotnet-new.md) or [Visual Studio](../../core/tutorials/with-visual-studio.md). Name it *InputTutorial*.

    ```dotnetcli
    dotnet new console -o InputTutorial
    ```

> 1. [!INCLUDE [tutorial-add-packages](../includes/tutorial-add-packages.md)]

Reviewer: I'm pretty sure we're missing a step here to `cd` into the newly created _InputTutorial_ project directory. Consider adding this missing step.

    :::code language="csharp" source="~/iot-samples/tutorials/InputTutorial/Program.cs" :::

    In the preceding code:

    - A [using declaration](../../csharp/whats-new/csharp-8.md#using-declarations) creates an instance of `GpioController`. The `using` declaration ensures the object is disposed and hardware resources are released properly.
    - GPIO pin 21 is opened with  `PinMode.InputPullUp`.
        - This opens the pin with a *PullUp* resistor engaged. In this mode, when the pin is connected to ground, it will return `PinValue.Low`. When the pin is disconnected from ground and the circuit is open, the pin returns `PinValue.High`.
    - The initial status is written to a console using a ternary expression. The pin's current state is read with `Read()`. If it's `PinValue.High`, it writes the `alert` string to the console. Otherwise, it writes the `ready` string.
    - `RegisterCallbackForPinValueChangedEvent()` registers an anonymous callback function for both the `PinEventTypes.Rising` and `PinEventTypes.Falling` events on the pin. These events correspond to pin states of `PinValue.High` and `PinValue.Low`, respectively.

>     - The callback function delegates to a method called `onPinEvent`. `onPinEvent` uses another ternary expression that also writes the corresponding `alert` or `ready` strings.

Reviewer: ```suggestion
    - The callback function delegates to a method called `OnPinEvent`. `OnPinEvent` uses another ternary expression that also writes the corresponding `alert` or `ready` strings.
```

Consider renaming this as all other C# methods.

Author: Wait... C# method names should be Pascal Case? 

Why have I been using Camel Case? My entire life is a lie.

Author: No, it turns out, I knew that. Then why does it suddenly seem wrong?

I'm scared. Hold me.

    - The main thread sleeps indefinitely while waiting for pin events.

1. [!INCLUDE [tutorial-build](../includes/tutorial-build.md)]
1. [!INCLUDE [tutorial-deploy](../includes/tutorial-deploy.md)]
1. Run the app on the Raspberry Pi by switching to the deployment directory and running the executable.

    ```bash
    ./InputTutorial
    ```

    The app displays text similar to the following:

    ```console

>     Initial status (05/10/2022 15:59:25): Ready ✅

Reviewer: ```suggestion
    Initial status (05/10/2022 15:59:25): READY ✅
```

    ```

1. Disconnect pin 21 from ground. The console displays text similar to the following:

    ```console

>     (05/10/2022 15:59:59) ALERT! 🚨

Reviewer: ```suggestion
    (05/10/2022 15:59:59) ALERT 🚨
```

    ```

1. Reconnect pin 21 and ground. The console displays text similar to the following:

    ```console

>     (05/10/2022 16:00:25) Ready ✅

Reviewer: ```suggestion
    (05/10/2022 16:00:25) READY ✅
```

    ```

1. Terminate the program by pressing <kbd>Ctrl</kbd>+<kbd>C</kbd>.

> Congratulations! You've used GPIO to detect input! There are many uses for this type of input. This example can be used with any scenario where a switch connects or breaks a circuit. Here's an example using it with a magnetic reed switch, which is often used to detect open doors or windows.

Reviewer: ```suggestion
Congratulations, you've used GPIO to detect input using the `Iot.Device.Bindings` NuGet package! There are many uses for this type of input. This example can be used with any scenario where a switch connects or breaks a circuit. Here's an example using it with a magnetic reed switch, which is often used to detect open doors or windows.
```

:::image type="content" source="https://via.placeholder.com/600x400.png" alt-text="Animated GIF showing magnetic reed switch":::

## Laser tripwire

> A similar application involves laser tripwires. Building a laser tripwire requires the following additional components:

Reviewer: Let's transition a bit from the previous example.

```suggestion
Extending the previous example concept a bit further, let's take a look at how this could be applied to creating a laser tripwire. Building a laser tripwire requires the following additional components:
```

* KY-008 laser transmitter module
* Laser receiver sensor module *(see note below)*
* 10K Ω resistor
* 20K Ω resistor

> [!NOTE]
> *Laser receiver sensor module* is the generic name applied to a common module found at many internet retailers. The device may vary in name or manufacturer, but should resemble this image.
>
> ![Image of a Laser Receiver Sensor Module](https://via.placeholder.com/200x200.png)

> ### Hardware

Reviewer: This heading is a bit generic, how about this instead?
```suggestion
### Connect laser tripwire hardware
```

Connect the components as detailed in following diagram.

:::image type="content" source="https://via.placeholder.com/500x300.png" alt-text="Fritzing diagram":::

> Pay special attention to the 10K Ω and 20K Ω resistors. These implement a [voltage divider](https://en.wikipedia.org/wiki/Voltage_divider). This is because the laser receiver module outputs 5V to indicate the beam is broken. Raspberry Pi only supports up to 3.3V for GPIO input. Since sending the full 5V to the pin could damage the Raspberry Pi, the current from the receiver module is passed through a voltage divider to halve the voltage to 2.5V.

Reviewer: ```suggestion
Pay close attention to the 10K Ω and 20K Ω resistors. These implement a [voltage divider](https://en.wikipedia.org/wiki/Voltage_divider). This is because the laser receiver module outputs 5V to indicate the beam is broken. Raspberry Pi only supports up to 3.3V for GPIO input. Since sending the full 5V to the pin could damage the Raspberry Pi, the current from the receiver module is passed through a voltage divider to halve the voltage to 2.5V.
```

Also, I heard a few peers mention in the past that we should try to avoid Wikipedia links as the validity is unverified. Instead, consider creating a .NET IoT glossary article, and writing your own interpretation of these sorts of things.

> ### Software changes

Reviewer: ```suggestion
### Apply source code updates
```

You can *almost* use the same code as earlier, with one exception. In the other examples, we used `PinMode.InputPullUp` so that when the pin is disconnected from ground and the circuit is open, the pin returns `PinValue.High`.

> However, in the case of the laser receiver module, we're not detecting an open circuit. Instead, we want the pin to act as a sink for current coming from the laser receiver module. In this case, we'll open the pin with `PinMode.InputPullDown`. This way, the pin returns `PinValue.Low` when it's getting no current, and `PinValue.High` when it receives current from the laser receiver module.

Reviewer: Consider showing the final version of the code for the laser tripwire.

ms.date: 10/04/2022
ms.topic: tutorial
ms.prod: dotnet
recommendations: false
---

# Use GPIO for binary input

General-purpose I/O (GPIO) pins can be configured to receive electrical signals as input. At its most basic level, this is useful for scenarios that detect the opening/closing of a circuit. Such circuits might include push buttons, toggle switches, reed switches, pressure switches, and other devices that represent binary (on/off) values by completing a circuit.

In this tutorial, you'll use .NET and your Raspberry Pi's GPIO pins to detect the opening and closing of a circuit.

## Prerequisites

- [!INCLUDE [prereq-rpi](../includes/prereq-rpi.md)]
- Jumper wires
- Breadboard (optional)
- Raspberry Pi GPIO breakout board (optional)
- [!INCLUDE [tutorial-prereq-dotnet](../includes/tutorial-prereq-dotnet.md)]

[!INCLUDE [ensure-ssh](../includes/ensure-ssh.md)]

## Prepare the hardware

Use the hardware components to build the circuit as depicted in the following diagram:

:::image type="content" source="../media/rpi-jumper-21-gnd-thumb.png" alt-text="A diagram showing a circuit that connects a ground pin to pin 21." lightbox="../media/rpi-jumper-21-gnd-thumb.png":::

> The image above depicts a direct connection between a ground pin and pin 21.

Reviewer 2: Is it ok to assume people would know which pin is pin 21? I'm not sure how many people are used to the Gpio numbering, there are also different numbering schemes to be aware of, like the board pin numbering or the logical GPIO one.

### Apply source code updates

You can *almost* use the same code as earlier, with one exception. In the other examples, we used `PinMode.InputPullUp` so that when the pin is disconnected from ground and the circuit is open, the pin returns `PinValue.High`.

However, in the case of the laser receiver module, we're not detecting an open circuit. Instead, we want the pin to act as a sink for current coming from the laser receiver module. In this case, we'll open the pin with `PinMode.InputPullDown`. This way, the pin returns `PinValue.Low` when it's getting no current, and `PinValue.High` when it receives current from the laser receiver module.

```csharp
controller.OpenPin(pin, PinMode.InputPullDown);
```

> [!IMPORTANT]
> Make sure the code deployed on your Raspberry Pi includes this change before testing a laser tripwire. The program *does* work without it, but using the wrong input mode risks damage to your Raspberry Pi!

:::image type="content" source="https://via.placeholder.com/600x400.png" alt-text="Animated GIF showing the laser tripwire":::

## Get the source code

> The source for this tutorial is [available on GitHub](https://github.com/MicrosoftDocs/dotnet-iot-assets/tree/main/tutorials/InputTutorial).

Reviewer 3: This area is where I'd give readers some ideas for how to apply it. Detecting a switch can do a lot of different things. Just some brainstorming:

- turning on indicators / displays.
- outside lighting / auto dusk-dawn turn on / turn off
- do not disturb
- bring coffee...

The laser is one of those ideas.
